File: 2sem-met1/matlab/test_met1.m
close all
clear all
clc
format long

%%
student = 0; % here is your number

N = student;
A = [
    4,  1,      1
    1,  6+.2*N, -1
    1,  -1,     8+.2*N
    ];
b = [1, -2, 3]';
x0 = [0, 0, 0]';
eps = 1e-6;

methods = {'met1_mngs', 'met1_mps'};

styles = {'mo-', 'b.:'};

hold on
for i = 1:numel(methods)
    method = methods{i};
    disp(['running ', method, ' test for student #', num2str(student)])
    [X, Y] = feval(method, A, b, x0, eps);
    
    x1 = linsolve(A, -b);
    y1 = 1/2*x1'*A*x1 + b'*x1;
    
    assert(norm(x1 - X(:, end)) < 1e-3)
    assert(norm(y1 - Y(:, end)) < eps)
    plot(-log10(Y - y1), styles{i})
end
xlabel('номер итерации')
ylabel('точность')
legend show

File: 2sem-met1/py/test_met1.py
import pytest
import numpy as np
import matplotlib.pyplot as plt
import met1

default_student = 0


@pytest.mark.parametrize("student", [default_student])  # student num
@pytest.mark.parametrize("n_dim", [2, 3])               # dimension number
def test_met1(student, n_dim):
    print(f'running met1 test for student #{student}')
    N = student

    A = np.array([[4,  1,      1],
                  [1,  6+.2*N, -1],
                  [1,  -1,      8+.2*N]],
                 dtype='float')[:n_dim, :n_dim]
    b = np.array([1, -2, 3], dtype='float').reshape(-1, 1)[:n_dim, :]
    x0 = np.array([0, 0, 0], dtype='float').reshape(-1, 1)[:n_dim, :]

    eps = 1e-6

    methods = ['mngs', 'mps']
    styles = ['mo-', 'b.:']
    plt.figure()
    plt.xlabel('номер итерации')
    plt.ylabel('точность')
    for i, method in enumerate(methods):
        X, Y = getattr(met1, method)(A, b, x0, eps)

        x1 = np.linalg.solve(A, -b)
        y1 = (1/2 * x1.T @ A @ x1 + b.T @ x1).item()

        assert np.equal(x0, X[0]).all()
        assert np.linalg.norm(x1 - X[-1]) < 1e-3
        assert np.linalg.norm(y1 - Y[-1]) < eps

        plt.plot(-np.log10([y - y1 for y in Y]), styles[i])
    plt.legend(methods)
    plt.show()
